### PR TITLE
Don't re-encode byte[] values in SortValues transform

### PR DESCRIPTION
**Pull request by clairemcginty**
A small optimization for `SortValues` transform to avoid doubly roundtrip encoding values that are already `byte[]`s -- they can be passed directly to the Sorter.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/SortValues.java [hunk 1]
```diff
@@ -167,6 +170,55 @@ public void processElement(ProcessContext c) {
       }
     }
 
+    @FunctionalInterface
+    private interface ThrowingFunction<T, U> {
+      U apply(T t) throws IOException, CoderException;
+    }
+
+    private class KvByteTranslator {
```
Comment (clairemcginty) [inline sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/SortValues.java:178]: This feels like a particularly inelegant solution but I didn't have a better idea 😅

**Comment by Abacn** [comment]
thanks, will take a look

**Review by damondouglas — CHANGES_REQUESTED** [review]
Thank you for contributing this!  The following three proposals go together and made more sense to provide in one comment, rather than separately. May we consider instead of the new custom class:
1. two private static methods as shown below:
```java
private static <T> T elementOf(Coder<T> coder, byte[] bytes) throws CoderException {
    if (coder instanceof ByteArrayCoder) {
      return (T) bytes;
    }
    return CoderUtils.decodeFromByteArray(coder, bytes);
  }

  private static <T> byte[] bytesOf(Coder<T> coder, T element) throws CoderException {
    if (element instanceof byte[]) {
      return (byte[]) element;
    }
    return CoderUtils.encodeToByteArray(coder, element);
  }
```
2. Refactoring the SortValuesDoFn's processElement method with:

```java
for (KV<SecondaryKeyT, ValueT> record : records) {
    sorter.add(KV.of(
        bytesOf(keyCoder, record.getKey()),
        bytesOf(valueCoder, record.getValue()))
    );
}
```

3. Refactoring DecodingIterator's next with:

```java
SecondaryKeyT secondaryKey = elementOf(keyCoder, next.getKey());
ValueT value = elementOf(valueCoder, next.getValue());
return KV.of(secondaryKey, value);
```

**Comment by clairemcginty** [comment]
> Thank you for contributing this! The following three proposals go together and made more sense to provide in one comment, rather than separately. May we consider instead of the new custom class:
> 
> 1. two private static methods as shown below:
> 
> ```java
> private static <T> T elementOf(Coder<T> coder, byte[] bytes) throws CoderException {
>     if (coder instanceof ByteArrayCoder) {
>       return (T) bytes;
>     }
>     return CoderUtils.decodeFromByteArray(coder, bytes);
>   }
> 
>   private static <T> byte[] bytesOf(Coder<T> coder, T element) throws CoderException {
>     if (element instanceof byte[]) {
>       return (byte[]) element;
>     }
>     return CoderUtils.encodeToByteArray(coder, element);
>   }
> ```
> 
> 2. Refactoring the SortValuesDoFn's processElement method with:
> 
> ```java
> for (KV<SecondaryKeyT, ValueT> record : records) {
>     sorter.add(KV.of(
>         bytesOf(keyCoder, record.getKey()),
>         bytesOf(valueCoder, record.getValue()))
>     );
> }
> ```
> 
> 3. Refactoring DecodingIterator's next with:
> 
> ```java
> SecondaryKeyT secondaryKey = elementOf(keyCoder, next.getKey());
> ValueT value = elementOf(valueCoder, next.getValue());
> return KV.of(secondaryKey, value);
> ```

Thanks for the review @damondouglas ! Sorry for the late reply, I've been out of office, but will apply these suggestions today or tomorrow 👍

**Comment by damondouglas** [comment]
@clairemcginty 
> Updated! your suggestions are much clearer -- my original approach constructed the custom functions up-front so that we wouldn't have to do the type-checking on every individual element, but I think the readability/clarity was just not worth it.

Thank you so much for listening! Apologies as I was taking a week off last week and will review now.

**Review by damondouglas — APPROVED** [review]
@clairemcginty LGTM. Thank you so much for your contribution! I will merge.

**Pull request merged**